src/libraries/System.Runtime.Serialization.BinaryFormat/src/PayloadReader.cs

```diff
@@ -0,0 +1,285 @@
+    /// <exception cref="SerializationException">When reading input from <paramref name="payload"/> encounters invalid Binary Format data.</exception>
+    /// <exception cref="DecoderFallbackException">When reading input from <paramref name="payload"/>
+    /// encounters invalid sequence of UTF8 characters.</exception>
+    public static SerializationRecord Read(Stream payload, PayloadOptions? options = default, bool leaveOpen = false)
+        => Read(payload, out _, options, leaveOpen);
+
+    /// <param name="payload">The Binary Format payload.</param>
+    /// <param name="recordMap">Record map</param>
+    /// <param name="options">An object that describes optional <seealso cref="PayloadOptions"/> parameters to use.</param>
+    /// <param name="leaveOpen">True to leave the <paramref name="payload"/> payload open
+    /// after the reading is finished, otherwise, false.</param>
+    /// <inheritdoc cref="Read(Stream, PayloadOptions?, bool)"/>
+    public static SerializationRecord Read(Stream payload, out IReadOnlyDictionary<int, SerializationRecord> recordMap, PayloadOptions? options = default, bool leaveOpen = false)
+    {
+#if NETCOREAPP
+        ArgumentNullException.ThrowIfNull(payload);
+#else
+        if (payload is null) throw new ArgumentNullException(nameof(payload));
+#endif
+
+        using BinaryReader reader = new(payload, ThrowOnInvalidUtf8Encoding, leaveOpen: leaveOpen);
+        return Read(reader, options ?? new(), out recordMap);
+    }
+
+    /// <summary>
+    /// Reads the provided Binary Format payload that is expected to contain an instance of any class (or struct) that is not an <seealso cref="Array"/> or a primitive type.
+    /// </summary>
+    /// <returns>A <seealso cref="ClassRecord"/> that represents the root object.</returns>
+    /// <inheritdoc cref="Read(Stream, PayloadOptions?, bool)"/>
+    public static ClassRecord ReadClassRecord(Stream payload, PayloadOptions? options = default, bool leaveOpen = false)
+        => (ClassRecord)Read(payload, options, leaveOpen);
+
+    private static SerializationRecord Read(BinaryReader reader, PayloadOptions options, out IReadOnlyDictionary<int, SerializationRecord> readOnlyRecordMap)
+    {
+        Stack<NextInfo> readStack = new();
+        var recordMap = new RecordMap();
+
+        // Everything has to start with a header
+        var header = (SerializedStreamHeaderRecord)ReadNext(reader, recordMap, AllowedRecordTypes.SerializedStreamHeader, options, out _);
+        // and can be followed by any Object, BinaryLibrary and a MessageEnd.
+        const AllowedRecordTypes allowed = AllowedRecordTypes.AnyObject
+            | AllowedRecordTypes.BinaryLibrary | AllowedRecordTypes.MessageEnd;
+
+        RecordType recordType;
+        SerializationRecord nextRecord;
+        do
+        {
+            while (readStack.Count > 0)
+            {
+                NextInfo nextInfo = readStack.Pop();
```

Comment:
For core this can be simplified.  And you could make it an extension method for polyfill for .NET Standard/.NET Framework concerns.

```suggestion
            while (readStack.TryPop(out NextInfo nextInfo)
            {
```

Reply:
> For core this can be simplified.

I am aware of that. But adding the polyfill brings too much complexity for me and I prefer to use a syntax that works everywhere instead.